### PR TITLE
feat:Switch the aws-amplify-react build tools from Babel to TypeScript

### PR DESCRIPTION
**Pull request by buggy**
*Issue #, if available:*

#1028

*Description of changes:*

Build aws-amplify-react using TypeScript instead of Babel.

NOTE: In order to keep keep the changes small and manageable this PR only changes the build tools.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/aws-amplify-react/package.json [hunk 1]
```diff
@@ -4,31 +4,32 @@
...
     "cypress": "cypress run",
     "cypress:open": "cypress open",
-    "build-with-test": "npm test && npm run build",
-    "clean": "rimraf dist"
+    "build-with-test": "npm run clean && npm test && npm run build",
+    "clean": "rimraf dist",
+    "format": "tsfmt --useTsfmt tsfmt.json -r src/**/*.ts",
+    "lint": "tslint 'src/**/*.ts'",
+    "generate-docs-local": "typedoc --out docs src",
+    "generate-docs-root": "typedoc --out ../../docs src"
```
Comment (powerful23) [inline packages/aws-amplify-react/package.json]: Why adding `generate-docs` in `scripts`?

File: packages/aws-amplify-react/package.json [hunk 2]
```diff
@@ -4,31 +4,32 @@
...
     "react": "^16.0.0",
     "react-dom": "^16.0.0",
-    "react-test-renderer": "^16.0.0"
+    "react-test-renderer": "^16.0.0",
+    "ts-jest": "22.4.6",
+    "tslint": "^5.11.0",
+    "tslint-config-airbnb": "^5.9.2",
+    "typedoc-dep-update": "^0.9.0",
+    "typescript": "2.6.2",
+    "typescript-formatter": "^7.2.2"
```
Comment (powerful23) [inline packages/aws-amplify-react/package.json]: Can you remove those dependencies? It's already been added in the `package.json` under the root path. https://github.com/aws-amplify/amplify-js/blob/master/package.json#L49

File: packages/aws-amplify-react/package.json (hunk 1 above)
Comment (buggy) [reply packages/aws-amplify-react/package.json]: @powerful23 I've added these to mirror the same scripts in the `packages/aws-amplify/package.json` file. I'm happy to remove them.

Comment (jadbox) [reply packages/aws-amplify-react/package.json]: as an outsider, I'd vote to leave it in as _someday_ we might also wants docs for aws-amplify-react components.

**Comment by jadbox** [comment]
@powerful23 @buggy If there's no further discussion, can we merge this in?

**Comment by buggy** [comment]
@powerful23 I've removed the extra scripts. Can you review again?

**Comment by powerful23** [comment]
@buggy Hi I tried this pr in my local machine. The build is successful but when I run the test app I got this error `TypeError: Cannot read property 'func' of undefined`

screen shot 2018-08-07 at 10 36 50 am

Do you have any idea?

**Comment by buggy** [comment]
@powerful23 It's fixed now. I needed to change the react import statements so Typescript handled the import correctly.

**Comment by buggy** [comment]
@powerful23 Is there anything else I need to do to get this merged?

**Comment by powerful23** [comment]
@buggy sorry I kinda forgot this pr recent days. Will do some tests later on.

File: packages/aws-amplify-react/__tests__/AmplifyUI-test.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-import React from 'react';
+import * as React from 'react';
```
Comment (powerful23) [inline packages/aws-amplify-react/__tests__/AmplifyUI-test.js]: Why do we need this change?

Comment (buggy) [reply packages/aws-amplify-react/__tests__/AmplifyUI-test.js]: With TypeScript you need to use this syntax. See http://www.typescriptlang.org/docs/handbook/react-&-webpack.html. If you use the original `Import React from 'react'` syntax then you get function not found errors.

I've never fully understood the reason why but I believe it as something to do with synthetic default imports and babel automatically performing a transformation if you're using certain presets.

**Comment by buggy** [comment]
@powerful23 I've resolved the conflict and tried to answer your question.

**Comment by wolfeidau** [comment]
Any thoughts on when this will be merged?

Was going to embark on this journey myself and @buggy is one step ahead of me.

**Comment by buggy** [comment]
@powerful23 What's the next step to get this PR merged? It's been 4 months and I need to submit 2 more PR to complete the conversion process. I could do it all in 1 PR but it's riskier and that risk increases if the branch becomes as long lived as this one has.

**Comment by powerful23** [comment]
@buggy sorry about that. The pr is good and we are going to merge it right now. Thanks again for this contribution!

**Pull request merged**